### ISSUE TITLE
about gradient

### ISSUE THREAD

**Issue by zzzhhhlll** [JonasGeiping/invertinggradients#14]
I would like to ask if it is possible to recover the trained image data if only the bias term data of the last layer of the model gradient is leaked

**Comment by JonasGeiping** [comment]
That depends on how wide the layer is. There is some info leaked, let's say the model is a classification model, then each value in the bias gradient is the change in this feature. For most models though, there would not be enough information present to reassemble the image data.

**Comment by zzzhhhlll** [comment]
> That depends on how wide the layer is. There is some info leaked, let's say the model is a classification model, then each value in the bias gradient is the change in this feature. For most models though, there would not be enough information present to reassemble the image data.

Thank you for your answer. I am using convolutional neural networks for image recognition. I would like to encrypt all gradients except for the last layer, so that only the last layer of information is leaked, and the width of the last layer is 10. This is not a big problem for privacy protection, is it

**Comment by JonasGeiping** [comment]
Assuming your images have more than 10 pixels, the recovery is at least underdetermined ;). Yeah, this seems fine.

**Comment by zzzhhhlll** [comment]
> Assuming your images have more than 10 pixels, the recovery is at least underdetermined ;). Yeah, this seems fine.

I use mnist and cifar10 datasets。 Only the last layer of plaintext of the gradient is leaked, so there should be no security issues, or is it uncertain

**Comment by JonasGeiping** [comment]
So the last layer, including both weights and biases? For a single image input, this would be sufficient to recover the features of the image. Whether that is security-critical, depends on your application. I can imagine scenarios where it would be possible to determine attribution to a small subpopulation based on the features online, but in most scenarios, this is likely unreasonable. You probably also use batch sizes greater than 1, which breaks such attacks in the average case, and only the average of the features of all images can be recovered.

**Comment by zzzhhhlll** [comment]
> So the last layer, including both weights and biases? For a single image input, this would be sufficient to recover the features of the image. Whether that is security-critical, depends on your application. I can imagine scenarios where it would be possible to determine attribution to a small subpopulation based on the features online, but in most scenarios, this is likely unreasonable. You probably also use batch sizes greater than 1, which breaks such attacks in the average case, and only the average of the features of all images can be recovered.

The last layer I am referring to only has bias,
eg  
fc1weight  (full connect)
fc1bias
fc2weight
fc2bias

I only leak fc2bias，Except for the last layer of bias, all other layers are in an encrypted state

**Comment by JonasGeiping** [comment]
Ah ok, then as I said, this seems pretty safe. There can be no absolute safety. For example, you could derive a state of parameters in the model that leads to a bias gradient that is high for only a single data point, but not for any others, and would allow identification of that data point. Yet, usual training would almost surely never lead to such a parameter state.

**Comment by zzzhhhlll** [comment]
> Ah ok, then as I said, this seems pretty safe. There can be no absolute safety. For example, you could derive a state of parameters in the model that leads to a bias gradient that is high for only a single data point, but not for any others, and would allow identification of that data point. Yet, usual training would almost surely never lead to such a parameter state.

ok,I have a rough understanding, so it is the safest state to not leak any layer of gradient. For gradient compression like that, top k, with a sparsity of 99%, only 1% of the important gradient is transmitted. Even if it is obtained, it should not be able to restore the image.

**Comment by zzzhhhlll** [comment]
> Ah ok, then as I said, this seems pretty safe. There can be no absolute safety. For example, you could derive a state of parameters in the model that leads to a bias gradient that is high for only a single data point, but not for any others, and would allow identification of that data point. Yet, usual training would almost surely never lead to such a parameter state.

Do you have any image restoration for the last layer of Bias leakage in your paper? Would you like to know more about it

**Comment by JonasGeiping** [comment]
Maybe I am being confusing here. All I'm saying is that from a worst-case perspective, an adversary could design a state of weights/parameters where that 1% of information could be able to identify a user. Identification (meaning "if I have the image, can I, as an attacker,  verify that the gradient I received is the result of the user having this image") is a strictly easier problem than image restoration, and a first step in privacy leakage. 


Image restoration seems unlikely. It's again hard for me to say impossible, because you could construct images where only one pixel is important, but for all practical purposes, no restoration could be done. 


EDIT: Robbing the Fed https://openreview.net/forum?id=fwzUgo0FM9v and Fishing for User Data https://proceedings.mlr.press/v162/wen22a.html are maybe interesting for you, even if neither discusses exactly the case where only the last layer is leaked.

**Comment by zzzhhhlll** [comment]
> Maybe I am being confusing here. All I'm saying is that from a worst-case perspective, an adversary could design a state of weights/parameters where that 1% of information could be able to identify a user. Identification (meaning "if I have the image, can I, as an attacker, verify that the gradient I received is the result of the user having this image") is a strictly easier problem than image restoration, and a first step in privacy leakage.
> 
> Image restoration seems unlikely. It's again hard for me to say impossible, because you could construct images where only one pixel is important, but for all practical purposes, no restoration could be done.
> 
> EDIT: Robbing the Fed https://openreview.net/forum?id=fwzUgo0FM9v and Fishing for User Data https://proceedings.mlr.press/v162/wen22a.html are maybe interesting for you, even if neither discusses exactly the case where only the last layer is leaked.

Thank you, a pleasant discussion.

**Comment by zzzhhhlll** [comment]
> Ah ok, then as I said, this seems pretty safe. There can be no absolute safety. For example, you could derive a state of parameters in the model that leads to a bias gradient that is high for only a single data point, but not for any others, and would allow identification of that data point. Yet, usual training would almost surely never lead to such a parameter state.

I also want to ask a question, which is more harmful, gradient leakage or model parameter leakage， I think there was a bit of a question I asked before. Do the general model parameters include gradients? Do I want to leak the last layer weight values of the model, and will this leakage restore the image? I have read the article "Deep Leakage from Gradients" before, which said that compressing gradients can prevent leakage. So, can I also shrink the model parameters to prevent leakage
